### PR TITLE
power of a ratio with a ratio exponent

### PR DESCRIPTION
**Pull request by gitsbi**

### CODE REVIEW

File: math.hpp [hunk 1]
```diff
@@ -12,21 +12,109 @@
...
+ * @tparam U  Quantity's unit
+ * @tparam S  Quantity's scale
+ * @tparam V  Quantity's value type
+ * @tparam T  Quantity's tag
+ *
+ * @param q  Quantity to get cube root from
+ *
+ * @return cube root of @p q
+ *
+ * @note The square root is only defined for selected values of @p S and @p T.
```
Comment (haphi) [inline math.hpp]:
```suggestion
 * @note The cube root is only defined for selected values of @p S and @p T.
```

File: ratio.hpp [hunk 2]
```diff
@@ -25,32 +26,135 @@ template< std::intmax_t Num
...
+template<>                               struct sign<0     ,false> { static const int value =  0; };
+
+template<typename Ratio, int Power, int Sign = sign<Power>::value>
+struct ratio_pow;
+
+template<typename Ratio, int Power>
+using ratio_pow_t = typename ratio_pow<Ratio,Power>::type;
+
+template<typename Ratio , int Sign > struct ratio_pow<     Ratio     ,     0, Sign> {using type = ratio_t<1>;};
+template<std::intmax_t D, int Power> struct ratio_pow<std::ratio<0,D>, Power,   -1> {using type = ratio_t<0>;};
```
Comment (haphi) [inline ratio.hpp]: this is `1/(0^P)`, which is undefined (or +inf, if you like). should probably fail a static assertion

File: scaling.hpp [hunk 3]
```diff
@@ -143,56 +143,18 @@ template<typename T> using to_day_scale    = typename detail::scale_to<   day_sc
...
  *
  * @brief Multiply and divide scale types
  */
 template<typename Scale1, typename Scale2> using  mul_scale_t = std::ratio_multiply<Scale1,Scale2>;
 template<typename Scale1, typename Scale2> using  div_scale_t = std::ratio_divide  <Scale1,Scale2>;
-template<typename Scale ,      int Power > using  pow_scale_t = typename detail::pow<Scale,Power>::type;
-template<typename Scale                  > using sqrt_scale_t = typename detail::sqrt<Scale>::type;
+template<typename Scale , typename Ratio > using  pow_scale_t = ratio_root_t< ratio_pow_t< Scale
+                                                                                         , abs_t<Ratio>::num >
+                                                                            , abs_t<Ratio>::den >;
```
Comment (haphi) [inline scaling.hpp:155]:
```suggestion
                                                                                         , abs_t<Ratio>::den >;
```

File: math.hpp [hunk 4]
```diff
@@ -12,21 +12,109 @@
...
+ * @param q  Quantity to get square root from
+ *
+ * @return square root of @p q
+ *
+ * @note The square root is only defined for selected values of @p S and @p T.
+ *       @see ratio_root_t for more details.
+ *
+ * @note This operation will affect @p U, @p S, and @p T as well. If the type
+ *       of @p q is not the result of a quantity raised to the power of two,
+ *       then the resulting scale will be an awkward approximation of the
```
Comment (haphi) [inline math.hpp]: To my relief, this is not true

File: math.hpp [hunk 5]
```diff
@@ -12,21 +12,109 @@
...
+ * @tparam     T  Quantity's tag
+ *
+ * @param q  Quantity to raise
+ *
+ * @return @p q, raised to Power
+ */
+template<typename Power, typename U, typename S, typename V, typename T>
+auto pow(const quantity<U,S,T,V>& q) {
+	return pow_quantity_t<quantity<U,S,T,V>, Power>{ std::pow( static_cast<double>(q.get())
+	                                                         , static_cast<double>(Power::num)/Power::den ) };
```
Comment (haphi) [inline math.hpp]: `sqrt` is often way faster than `pow(x,0.5)` (and supposedly more accurate as well, iirc ieee). So I would suggest specialization. anyway as a user, I would expect `sqrt(some_quantity)` to call `std::sqrt`under the hood, not  `std::pow`.

File: math.hpp [hunk 6]
```diff
@@ -12,21 +12,109 @@
...
+ * @param q  Quantity to get cube root from
+ *
+ * @return cube root of @p q
+ *
+ * @note The square root is only defined for selected values of @p S and @p T.
+ *       @see ratio_root_t for more details.
+ *
+ * @note This operation will affect @p U, @p S, and @p T as well. If the type
+ *       of @p q is not the result of a quantity raised to the power of three,
+ *       then the resulting scale will be an awkward approximation of the
```
Comment (haphi) [inline math.hpp]: not true, it will just not compile

File: test/test_ratio.cpp [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+
+#include <unlib/test/unlib_test.hpp>
+
+#include <unlib/scaling.hpp>
+
+namespace {
+
+using doctest::Approx;
+
+constexpr double get_float(std::intmax_t num, std::intmax_t den) {return static_cast<double>(num)/static_cast<double>(den);}
```
Comment (haphi) [inline test/test_ratio.cpp]: the name suggests `float` return value, why not `get_double`?

File: math.hpp (hunk 5 above)
Comment (haphi) [reply math.hpp]: quick experiment on my machine with a simple loop of 1 billion operations shows runtime with sqrt of `0.15s` (sqrt)  vs. `1.6s` (pow)

File: math.hpp (hunk 1 above)
Comment (gitsbi) [reply math.hpp]: _Done._

File: ratio.hpp (hunk 2 above)
Comment (gitsbi) [reply ratio.hpp]: _Removed._

File: scaling.hpp (hunk 3 above)
Comment (gitsbi) [reply scaling.hpp:155]: Unfortunately, github seems to mangle your suggestions, so I have no idea what you are trying to get across.

File: math.hpp (hunk 4 above)
Comment (gitsbi) [reply math.hpp]: _Text removed._

File: math.hpp (hunk 5 above)
Comment (gitsbi) [reply math.hpp]: Changing this for `pow<ratio_t<1,2>>()` took some effort, but I think I got it covered now.

I have verified that it works in the debugger, but I am not sure how to automatically test whether the correct std lib function is called.

File: math.hpp (hunk 6 above)
Comment (gitsbi) [reply math.hpp]: _Text removed._

File: test/test_ratio.cpp (hunk 7 above)
Comment (gitsbi) [reply test/test_ratio.cpp]: /shrug I needed a floating point number in a test. 

_Renamed._

**Pull request merged**